clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -365,28 +389,142 @@ synchronized void resetProducerId() {
+                        + " for partition " + batch.topicPartition + " is going to become negative :" + newSequence);
+
+            log.debug("Setting sequence number of batch with current sequence {} for partition {} to {}", batch.baseSequence(), batch.topicPartition, newSequence);
+            inFlightBatch.setProducerState(new ProducerIdAndEpoch(inFlightBatch.producerId(), inFlightBatch.producerEpoch()), newSequence, inFlightBatch.isTransactional());
+        }
+    }
+
+    synchronized boolean hasInflightBatches(TopicPartition topicPartition) {
+        return inflightBatchesBySequence.containsKey(topicPartition) && !inflightBatchesBySequence.get(topicPartition).isEmpty();
+    }
+
+    synchronized boolean partitionHasUnresolvedSequence(TopicPartition topicPartition) {
+        return partitionsWithUnresolvedSequences.contains(topicPartition);
+    }
+
+    synchronized void markSequenceUnresolved(TopicPartition topicPartition) {
+        partitionsWithUnresolvedSequences.add(topicPartition);
+    }
+
+    // Checks if there are any partitions with unresolved partitions which may now be resolved. Returns true if
+    // the producer id needs a reset, false otherwise.
+
+    synchronized boolean maybeResolveSequences() {
+        for (TopicPartition topicPartition : partitionsWithUnresolvedSequences) {
+            if (!hasInflightBatches(topicPartition)) {
+                // The partition has been fully drained. At this point, the last ack'd sequence should be once less than
+                // next sequence destined for the partition. If so, the partition is fully resolved. If not, we should
+                // reset the sequence number if necessary.
+                if (lastAckedSequence(topicPartition) == sequenceNumber(topicPartition) - 1) {
+                    // This would happen when a batch was expired, but subsequent batches succeeded.
+                    partitionsWithUnresolvedSequences.remove(topicPartition);
+                } else {
+                    // We would enter this branch if all in flight batches were ultimately expired in the producer.
+                    log.info("No inflight batches remaining for {}, last ack'd sequence for partition is {}, next sequence is {}. " +
+                            "Going to reset producer state.", topicPartition, lastAckedSequence(topicPartition), sequenceNumber(topicPartition));
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    synchronized boolean isNextSequence(TopicPartition topicPartition, int sequence) {
+        return (!lastAckedSequence.containsKey(topicPartition) && sequence == 0) ||
+                (lastAckedSequence.containsKey(topicPartition) && (sequence - lastAckedSequence.get(topicPartition) == 1));
+    }
+
+    synchronized  void setNextSequence(TopicPartition topicPartition, int sequence) {
```

Comment:
nit: this could be `private`? Also, there's an unneeded space after `synchronized`.